### PR TITLE
STOP and PAUSE in pipeline task context

### PR DESCRIPTION
**Pull request by lyakh**
Resurrecting the reverted #4724 - hopefully no regressions this time. Tested with an addition of https://github.com/thesofproject/sof/commit/714176252a91674d71a98d0d1c432c39ed4f0482 to pass multiple pipeline and multiple pause-resume on CNL with a multicore HDA topology.

### CODE REVIEW

**Comment by lyakh** [comment]
Both CI failures https://sof-ci.01.org/sofpr/PR4960/build11004/devicetest/?model=CML_RVP_SDW&testcase=check-suspend-resume-with-playback-5 and https://sof-ci.01.org/sofpr/PR4960/build11004/devicetest/?model=TGLH_RVP_HDA&testcase=check-suspend-resume-with-capture-5 are timeouts in suspend-resume tests, seem unrelated

**Comment by keyonjie** [comment]
Looks this doesn't fix the multi-core xrun issues according to test plan result 7973, but good news is that looks there is no significant regression with it.

I will need some explain about how the start/stop in the pipeline task context work, especially for cases that there are pipeline list connected with this kind of triggering. Will check detail next week.

**Comment by keyonjie** [comment]
Hi @lyakh do you have clear logic or documentation to demonstrate of how the trigger start/stop sequence looks like with your design, e.g. for below 0,0 and 0,3 PCM devices:
image

We have 3 pipelines and 2 PCM devices here, when the trigger start/stop of PCM 0,0 and 0,3 IPC came from the Linux host side, how those pipelines 1, 7, 11 are started/stopped?

File: src/audio/pipeline/pipeline-stream.c [hunk 1]
```diff
@@ -119,14 +111,22 @@ static int pipeline_comp_trigger(struct comp_dev *current,
 
 	/* send command to the component and update pipeline state */
 	err = comp_trigger(current, ppl_data->cmd);
+	current->pipeline->trigger.aborted = err == PPL_STATUS_PATH_STOP;
```
Comment (lgirdwood) [inline src/audio/pipeline/pipeline-stream.c]: this looks wrong.

File: src/include/sof/audio/pipeline.h [hunk 2]
```diff
@@ -77,6 +77,7 @@ struct pipeline {
 		int cmd;
 		struct comp_dev *host;
 		unsigned int delay;	/* period count */
+		bool aborted;
```
Comment (lgirdwood) [inline src/include/sof/audio/pipeline.h:80]: can we comment e.g. ```/* trigger aborted due to X, Y, Z */```

File: src/schedule/ll_schedule.c [hunk 3]
```diff
@@ -123,22 +123,31 @@ static void schedule_ll_tasks_execute(struct ll_schedule_data *sch)
...
 	struct list_item *wlist;
-	struct list_item *tlist;
 	struct task *task;
 
 	/* check each task in the list for pending */
-	list_for_item_safe(wlist, tlist, &sch->tasks) {
-		task = container_of(wlist, struct task, list);
+	wlist = sch->tasks.next;
+
+	while (wlist != &sch->tasks) {
```
Comment (lgirdwood) [inline src/schedule/ll_schedule.c:131]: what are we iteration over  - best to comment.

File: src/audio/pipeline/pipeline-schedule.c [hunk 4]
```diff
@@ -38,14 +38,68 @@ static void pipeline_schedule_cancel(struct pipeline *p)
...
+			/* No delay: the final stage has already run too */
+		} else if (p->status == COMP_STATE_PAUSED && !p->trigger.aborted) {
+			err = SOF_TASK_STATE_COMPLETED;
+		} else {
+			p->status = COMP_STATE_ACTIVE;
+			err = SOF_TASK_STATE_RESCHEDULE;
+		}
+	}
+
+	p->trigger.cmd = -EINVAL;
```
Comment (lgirdwood) [inline src/audio/pipeline/pipeline-schedule.c:88]: wont this always return an error ?

**Review by lgirdwood — COMMENTED** [review]
Not following some parts.

File: src/audio/pipeline/pipeline-stream.c (hunk 1 above)
Comment (lyakh) [reply src/audio/pipeline/pipeline-stream.c]: @lgirdwood sorry, could you explain? I don't see what's wrong with this - if one of components returns `PATH_STOP` we abort trigger processing?

File: src/audio/pipeline/pipeline-schedule.c (hunk 4 above)
Comment (lyakh) [reply src/audio/pipeline/pipeline-schedule.c:88]: why return an error? This just indicates, that currently there's no trigger command to be processed by the task. If one of `START`, `STOP`, `PAUSE`, `RELEASE` arrives, `trigger.cmd` will be set to it and the trigger will be executed

Comment (lgirdwood) [reply src/audio/pipeline/pipeline-schedule.c:88]: ok, can we have a comment here and return 0 (no error, but nothing todo ?)

**Comment by lyakh** [comment]
CI: a single suspend-resume device test https://sof-ci.01.org/sofpr/PR4960/build11067/devicetest/?model=CML_RVP_SDW&testcase=check-suspend-resume-with-playback-5 failed with a timeout, QB seems to have a problem with their ICL device - same failure as in #4983

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**